### PR TITLE
Add tests of `sample_relative` and fix type of return values of `SkoptSampler` and `PyCmaSampler`

### PR DESCRIPTION
**Pull request by toshihikoyanase**
## Motivation

Currently, tests of the `sample_relative` methods are missing in `tests/sampler_tests/test_sampler.py` while the `sample_independent` methods are tested in it. I think this is because only a few samplers implemented `sample_relative` when we introduced `sample_relative` and each sampler with `sample_relative` tested it in the individual test file (i.e., `tests/sampler_tests/test_cmaes.py`).

Now that multiple samplers implement `sample_relative`, I think we can introduce the test cases for `sample_relative` to `tests/sampler_tests/test_sampler.py`.

## Description of the changes

- Add test cases of `sample_relative` for `TPESampler(multivariate=True)`, `CmaEsSampler`, `SkoptSampler` and `PyCmaSampler`.
- Fix type of return values of `SkoptSampler` and `PyCmaSampler`.
  - This type checking can be seen in the test cases of `sample_independent`.

### CODE REVIEW

File: tests/samplers_tests/test_samplers.py [hunk 1]
```diff
@@ -225,6 +234,120 @@ def sample() -> float:
...
+            (
+                UniformDistribution,
+                LogUniformDistribution,
+                DiscreteUniformDistribution,
+                IntUniformDistribution,
+                IntLogUniformDistribution,
+            ),
+        )
+        assert value >= distribution.low
+        assert value <= distribution.high
```
Comment (himkt) [inline tests/samplers_tests/test_samplers.py:285]:
```suggestion
        assert distribution.low <= value <= distribution.high
```

[nits]
It is really nits but it aligns with `test_sample_relative_mixed`.

https://github.com/optuna/optuna/pull/2897/files#diff-8b85ccf8a3534b2aa878bce8ece8ce930af410ab178ebc6d80031badbabb5052R345

Comment (toshihikoyanase) [reply tests/samplers_tests/test_samplers.py]: Good catch! Thank you!

**Comment by hvy** [comment]
@nzw0301 could you review this PR? If you're busy, maybe @himkt could continue?

File: optuna/integration/cma.py [hunk 2]
```diff
@@ -459,8 +459,11 @@ def _to_optuna_params(
     ) -> Any:
 
         dist = search_space[param_name]
+        if isinstance(dist, UniformDistribution):
+            # Type of cma_param_value is np.floating, so cast it to Python's built-in float.
+            return float(cma_param_value)
         if isinstance(dist, LogUniformDistribution):
-            return math.exp(cma_param_value)
+            return float(math.exp(cma_param_value))
```
Comment (nzw0301) [inline optuna/integration/cma.py:466]: Maybe I'm missing something, but `math.exp` returns `float` value if its argument is `numpy.float` value, so `float` cast can be removed:

```suggestion
            return math.exp(cma_param_value)
```

I locally confirmed this as follows and also checked the `tests/samplers_tests/test_samplers.py` was passed.
```bash
python -c "import numpy as np; import math; print(isinstance(math.exp(np.float64(0.)), np.floating))"
False
```

**Review by nzw0301 — COMMENTED** [review]
I've left one minor comment. But I'll look at more in the next week. 

Have a good weekend 🍹

**Comment by nzw0301** [comment]
Probably this is out-of-scope of this PR, but let me ask one thing: Why isn't BoTorchSampler in `tests/samplers_tests/test_samplers.py`?

I locally confirm that adding the sampler to 

- `parametrize_multi_objective_sampler`: no test failed errors.
- `parametrize_sampler`: one failed case related to nan's error handling
- `parametrize_relative_sampler`: many failed cases.

**Review by hvy — COMMENTED** [review]
Thanks for the PR!
I left some comments and questions if you could take a look.

On a side note, it's a tiny bit worrisome that tests in `tests/samplers_tests/test_samplers.py` all quite varied in style (e.g. existing independent tests are using numpy arrays, variable naming conventions, etc.). Do you think we can or should address this perhaps in a different PR?

File: optuna/integration/cma.py (hunk 2 above)
Comment (toshihikoyanase) [reply optuna/integration/cma.py]: You're absolutely right. I fixed it in https://github.com/optuna/optuna/pull/2897/commits/e812161c4047f58c782ce849b101fe71acf0d726. Thank you.

File: tests/samplers_tests/test_samplers.py [hunk 3]
```diff
@@ -225,6 +234,119 @@ def sample() -> float:
...
+)
+def test_sample_relative_numerical(
+    relative_sampler_class: Callable[[], BaseSampler],
+    x_distribution: BaseDistribution,
+    y_distribution: BaseDistribution,
+) -> None:
+
+    search_space = {"x": x_distribution, "y": y_distribution}
+    study = optuna.study.create_study(sampler=optuna.samplers.RandomSampler(seed=1))
+    for _ in range(4):
```
[the remark this answers is not in the document]
Comment (toshihikoyanase) [reply tests/samplers_tests/test_samplers.py]: Thank you for pointing it out. I found that the number `1` is enough to determine the search space. I fixed it in https://github.com/optuna/optuna/pull/2897/commits/4c544e5e2d76a902710b7c4ebbe860eae33ceac6.

File: tests/samplers_tests/test_samplers.py [hunk 4]
```diff
@@ -225,6 +234,119 @@ def sample() -> float:
...
+    ],
+)
+def test_sample_relative_numerical(
+    relative_sampler_class: Callable[[], BaseSampler],
+    x_distribution: BaseDistribution,
+    y_distribution: BaseDistribution,
+) -> None:
+
+    search_space = {"x": x_distribution, "y": y_distribution}
+    study = optuna.study.create_study(sampler=optuna.samplers.RandomSampler(seed=1))
```
[the remark this answers is not in the document]
Comment (toshihikoyanase) [reply tests/samplers_tests/test_samplers.py]: I should use the parametrized samplers. I fixed it in https://github.com/optuna/optuna/pull/2897/commits/4c544e5e2d76a902710b7c4ebbe860eae33ceac6. Thanks!

File: tests/samplers_tests/test_samplers.py [hunk 5]
```diff
@@ -225,6 +234,119 @@ def sample() -> float:
...
+            (
+                UniformDistribution,
+                LogUniformDistribution,
+                DiscreteUniformDistribution,
+                IntUniformDistribution,
+                IntLogUniformDistribution,
+            ),
+        )
+        assert distribution.low <= value <= distribution.high
+        assert not isinstance(value, np.floating)
```
[the remark this answers is not in the document]
Comment (toshihikoyanase) [reply tests/samplers_tests/test_samplers.py]: I applied the suggestion in https://github.com/optuna/optuna/pull/2897/commits/b82c8c4d01b5ebae06b1dfc81eace38e71784652.

To distinguish `float` and `int`, I introduced conditional branches in the test case. If it makes test cases complicated, I'll split the test cases.

File: tests/samplers_tests/test_samplers.py [hunk 6]
```diff
@@ -225,6 +234,119 @@ def sample() -> float:
...
+) -> None:
+
+    search_space = {"x": x_distribution, "y": y_distribution}
+    study = optuna.study.create_study(sampler=optuna.samplers.RandomSampler(seed=1))
+    for _ in range(4):
+        trial = study.ask(search_space)
+        study.tell(trial, sum(trial.params.values()))
+    study.sampler = relative_sampler_class()
+
+    values = study.sampler.sample_relative(study, _create_new_trial(study), search_space)
```
[the remark this answers is not in the document]
Comment (toshihikoyanase) [reply tests/samplers_tests/test_samplers.py]: Thank you for your comments. It's important to keep the test style to keep the maintainability.

> all quite varied in style (e.g. existing independent tests are using numpy arrays, variable naming conventions, etc.)

I also agree with the comment, so I restructured the sample_relative tests and employed a similar style to the sample_independent tests in https://github.com/optuna/optuna/pull/2897/commits/4c544e5e2d76a902710b7c4ebbe860eae33ceac6.

File: tests/samplers_tests/test_samplers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hvy) [reply tests/samplers_tests/test_samplers.py]: I think it's acceptable, thanks!

File: tests/samplers_tests/test_samplers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hvy) [reply tests/samplers_tests/test_samplers.py]: Thanks!

File: tests/samplers_tests/test_samplers.py [hunk 7]
```diff
@@ -229,6 +242,142 @@ def sample() -> float:
...
+    x_distribution: BaseDistribution,
+    y_distribution: BaseDistribution,
+) -> None:
+
+    search_space: Dict[str, BaseDistribution] = OrderedDict(x=x_distribution, y=y_distribution)
+    study = optuna.study.create_study(sampler=relative_sampler_class())
+    trial = study.ask(search_space)
+    study.tell(trial, sum(trial.params.values()))
+
+    def sample() -> List[float]:
```
Comment (hvy) [inline tests/samplers_tests/test_samplers.py:277]: A nit but we're also parameterizing over int distributions.
```suggestion
    def sample() -> List[Union[int, float]]:
```

File: tests/samplers_tests/test_samplers.py [hunk 8]
```diff
@@ -229,6 +242,142 @@ def sample() -> float:
...
+            (
+                UniformDistribution,
+                LogUniformDistribution,
+                DiscreteUniformDistribution,
+                IntUniformDistribution,
+                IntLogUniformDistribution,
+            ),
+        )
+        assert np.all(points[:, i] >= distribution.low)
+        assert np.all(points[:, i] <= distribution.high)
```
Comment (hvy) [inline tests/samplers_tests/test_samplers.py:294]: To be exact, we should change `<=` to `<` for uniform and loguniform, but I also think this is acceptable.

**Review by hvy — CHANGES_REQUESTED** [review]
Thanks for the quick fixes, it looks more aligned with the existing tests now. It basically LGTM besides some minor comments and the comment from @nzw0301.

File: tests/samplers_tests/test_samplers.py (hunk 8 above)
Comment (toshihikoyanase) [reply tests/samplers_tests/test_samplers.py:294]: Thank you for your comment. #2223 changed the range definitions of `UniformDistribution` and `LogUniformDistribution`, and I use `<=` for those. (Honestly speaking, the definitions are different from that of `suggest_float` and it is confusing. We may want to revise it in the version 3 discussion.)

I checked the distribution reference and the current specification of the ranges is as follows:

- [`UniformDistribution`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.distributions.UniformDistribution.html#optuna.distributions.UniformDistribution): `[low , high]` (Both included)
- [`LogUniformDistribution`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.distributions.LogUniformDistribution.html#optuna.distributions.LogUniformDistribution): `[low, high]`
- [`DiscreteUniformDistribution`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.distributions.DiscreteUniformDistribution.html#optuna.distributions.DiscreteUniformDistribution): `[low, high]`
- [`IntUniformDistribution`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.distributions.IntUniformDistribution.html#optuna.distributions.IntUniformDistribution): `[low, high]`
- [`IntLogUniformDistribution`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.distributions.IntLogUniformDistribution.html#optuna.distributions.IntLogUniformDistribution): `[low, high]`

**Review by hvy — APPROVED** [review]
LGTM!

I think that the tests could be improved a bit to not depend on numpy but it's open for discussion and probably outside the scope of this PR. Same goes for the BoTorch sampler parameterization.

**Comment by toshihikoyanase** [comment]
> Probably this is out-of-scope of this PR, but let me ask one thing: Why isn't BoTorchSampler in `tests/samplers_tests/test_samplers.py`?
> 
> I locally confirm that adding the sampler to
> 
> * `parametrize_multi_objective_sampler`: no test failed errors.
> * `parametrize_sampler`: one failed case `test_sample_single_distribution`. `BotorchSampler` takes `nan` values
> * `parametrize_relative_sampler`: the introduced tests by this PR failed. This is because the tests uses `dict` to store distributions as a search space, but botorch assumes that `OrderedDict` in `sample_relative`: https://github.com/optuna/optuna/blob/20b3fa30750d96511cdd34a67f8a94a6127bb5b6/optuna/integration/botorch.py#L431

@nzw0301 Thank you for pointing it out and I'm sorry for the delayed response.
I needed to fix `BoTorchSampler` to resolve the error in the second item, so I create a new PR to fix it.
https://github.com/optuna/optuna/pull/2928

After merging it, I'll create a follow-up PR to fix the error in the third item.

**Review by nzw0301 — APPROVED** [review]
LGTM. Thanks!

**Comment by hvy** [comment]
Seems like it's been stale although it's been approved. Let me merge it.

**Pull request merged**